python/revng/cli/_commands/daemon_self_test/__init__.py

```diff
@@ -0,0 +1,93 @@
+#
+# This file is distributed under the MIT License. See LICENSE.md for details.
+#
+
+import asyncio
+import os
+import sys
+from argparse import ArgumentParser
+from tempfile import TemporaryDirectory
+
+from revng.cli.commands_registry import Command, CommandsRegistry, Options
+from revng.cli.support import get_driver_options
+
+from .daemon_handler import DaemonProto, ExternalDaemonProto, InteractiveDaemonProto
+from .daemon_handler import NonInteractiveDaemonProto
+from .test import run_self_test
+
+
+class DaemonSelfTestCommand(Command):
+    FILTER_ENV = [
+        "STARLETTE_DEBUG",
+        "REVNG_NOTIFY_FIFOS",
+        "REVNG_ORIGINS",
+        "REVNG_DATA_DIR",
+        "REVNG_PROJECT_ID",
+    ]
+
+    def __init__(self):
+        super().__init__(
+            ("daemon-self-test",),
+            "Check if revng-daemon can produce all artifacts given an executable",
+        )
+
+    def register_arguments(self, parser: ArgumentParser):
+        parser.add_argument(
+            "-e",
+            "--external",
+            help="Use the specified external address instead of using the local daemon",
+        )
+        parser.add_argument(
+            "--revng-c",
+            action="store_true",
+            help="Enable additional checks, which assume revng-c is installed",
+        )
+        parser.add_argument(
+            "executable", metavar="EXECUTABLE", help="Executable to run the test with"
+        )
+
+    def run(self, options: Options):
+        args = options.parsed_args
+
+        if args.external is not None:
+            url = args.external
+            self.proto: DaemonProto = ExternalDaemonProto(url)
+        else:
+            temp_folder = TemporaryDirectory()
+            url = f"unix:{temp_folder.name}/daemon.sock"
+
+            interactive = args.gdb or args.lldb
+            driver_options = get_driver_options()
+            env = {k: v for k, v in os.environ.items() if k not in self.FILTER_ENV}
+
+            if interactive:
+                self.proto = InteractiveDaemonProto(url, driver_options, env)
+            else:
+                self.proto = NonInteractiveDaemonProto(url, driver_options, env)
```

Comment:
Managed to slim down the two handlers (code-wise) a bit, but IMHO they should remain separate.
Main differences:
* In non-interactive mode the output is captured and displayed only on `rc != 0`
* In non-interactive mode there are timeouts in place

Reply:
I still see 3 implementations, external, interactive, non-interactive.

Detecting `gdb` and `lldb` is hackish and not very future proof. Please let's not do it, let's keep things transparent, as they were designed to be.

The responsibility of timing out should be of who invokes the tests, not of the test itself.

Reply:
Time outs is a property of the test environment (e.g., `ctest`), not the test itself.

The deeper timeouts are hardcoded, the worse. No timeouts here, we can have them at higher levels.

Reply:
If we add timeouts at the test environment level we'll get back to the situation that happened a couple of months ago where the test run standalone would not fail but when run in the test suite would fail because of the cpu load of the other tests. If we add a long timeout (e.g. 30m) then everybody will complain if it happens to hang.

Reply:
I think we can do without timeouts.
All tests can potentially run indefinitely, this one is not special.

No hardcoded timeouts please.